### ISSUE TITLE
Identify similar patches to patch under development/submission

### ISSUE THREAD

**Issue by bulwahn** [lfd/PaStA#14]
Motivation:
Regularly, multiple developers, especially kernel janitors, address the same code issues concurrently without knowledge of the work of each other. So, some patches, especially cleanup patches, are developed and submitted to the mailing list that have been already addressed in very similar way by another patch submitted to the mailing list.

Task: Provide an interface for developers to identify similar patches to the patches they have under development, they consider for submission or that they have submitted to the mailing list.

For patches under development or considered for submission, this interface shall be a simple web-based interface (as online webservice) and a command-line tool (that should synchronize data when connected, but also work without internet connection).

For patches that are submitted, the interface should be an extension to patchwork and an email bot service for developers that opt-in to that service.

**Comment by bulwahn** [comment]
In the description, there is currently some overlap with the considered extension in https://github.com/lfd/PaStA/issues/9.

**Comment by vaniisgh** [comment]
Sorry about the delay in my follow up.
I wanted to confirm my understanding of this particular issue before I start work :)

>Task: Provide an interface for developers to identify similar patches to the patches they have under development, they consider for submission or that they have submitted to the mailing list.
For patches under development or considered for submission, this interface shall be a simple web-based interface (as online webservice) and a command-line tool (that should synchronize data when connected, but also work without internet connection).

I think this the the part with the overlap :) and I see that there's a WIP PR on that already and quite a bit of work going on that looks really cool.

What I understand is that there needs to be some sort of cli command that can be run on in the git repository that iteratively runs though the commits and in case of a mailing list bot should also go though the comments or replies.

Basically It should run the current analysis that pypasta runs just as a bot and cli tool.

For the cli tool I think it would be possible to use [argparse](https://docs.python.org/3/library/argparse.html) to run pypasta in the patch stack mode?

I am not sure how a mailbot works and will look into it too possibly.

I think if my understanding is close to what is needed I this might be a great starter issue for me to understand the current state of PaStA and the analysis components that run currently before attempting to add any further mathematics to it :)

**Comment by bulwahn** [comment]
@vaniisgh I do not understand anything you stated, it is either trivial unrelated stuff or just confusing. I will just ignore that.

**Comment by bulwahn** [comment]
@rralf I will try to describe here more what I have in mind:

On some server, we would have a pasta instance running that continuously:

1. updates all of its resources, picking all the latest emails from the public inbox archives.

2. Runs the typical pasta workflow, stores and prepares all patches for computing similarities (a differential analysis will probably  help here to save compute power)

3. Waits for client requests to provide a patch

At the client side, we have a small script that 

4. just sends a patch file that somebody wants to submit to this server.

5. The server then takes the patch file from every client, determines similar patches and returns all similar patches (or their lore.kernel.org links)

6. The client presents those similar patches.

7. The user of the client can then make sense of those responses. e.g., he can see if someone worked on roughly the same.

**Comment by rralf** [comment]
Ok, I think I understand. You want to answer the question: "Is my patch already addressed by someone else and present on some list?"

The first problem is, that you first have to write the patch before we can even answer the question, which probably requires most effort.

The second problem is, that it is very likely that different authors will write different commit message. We can solve this, as we can tell pasta to ignore the commit message for the comparison.

The third problem is, that we would have to compare the patch against a huge amount of patches. We can simplify this by only comparing the patch against patches within the last ~few months.

The fourth problem is that we need improvements in the pasta infrastructure to efficiently solve this issue. E.g., you don't want to start a new PaStA instance for every incoming mail. It takes minutes until all caches and list indices are loaded, which wastes time. This is the primary issue that we should work on. I might get a student that will work on a Client/Server model of PaStA: You have "pastad" running in background, keeping everything in memory during runtime. Then, there's a PaStA frontent that queue tasks. This will probably be helpful to solve your issue. In addition to that, the student will also consider to rework the storage backend by using a proper database as backend (Redis, SQL, whatever). But it's still uncertain if he'll pick up this task.

**Comment by vaniisgh** [comment]
thanks for taking the time to go through my comment, I think maybe I couldn't communicate my question properly along with the dace that my understanding is novice. I thought this issue was about creating CLI commands to run pasta when the user wants to run analyze their patches using it and having a mail bot that does the same. 

I don't know why I thought I user would want to do this on their own patches ... sorry about that.

Your description explains it a lot better, and it seems this is about running PaStA on all patches sent, by anyone. right?
